File: lib/personalize.js
import Groq from "groq-sdk";
import { consoleProgress } from "./progress.js";

const MODEL = "llama-3.3-70b-versatile";
const FETCH_TIMEOUT_MS = 8000;
const MAX_SIGNAL_CHARS = 1800;

async function fetchHtml(url) {
  const ctrl = new AbortController();
  const t = setTimeout(() => ctrl.abort(), FETCH_TIMEOUT_MS);
  try {
    const res = await fetch(url, {
      signal: ctrl.signal,
      redirect: "follow",
      headers: {
        "User-Agent":
          "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36",
        "Accept-Language": "pt-BR,pt;q=0.9,en;q=0.5",
      },
    });
    if (!res.ok) return "";
    const ct = res.headers.get("content-type") || "";
    if (!ct.includes("text") && !ct.includes("html")) return "";
    return await res.text();
  } catch {
    return "";
  } finally {
    clearTimeout(t);
  }
}

function stripTags(s) {
  return s
    .replace(/<script[\s\S]*?<\/script>/gi, " ")
    .replace(/<style[\s\S]*?<\/style>/gi, " ")
    .replace(/<[^>]+>/g, " ")
    .replace(/&nbsp;/gi, " ")
    .replace(/&amp;/gi, "&")
    .replace(/\s+/g, " ")
    .trim();
}

function extractSignals(html) {
  const title = html.match(/<title[^>]*>([\s\S]*?)<\/title>/i)?.[1] || "";
  const metaDesc =
    html.match(
      /<meta[^>]+name=["']description["'][^>]+content=["']([^"']+)["']/i,
    )?.[1] ||
    html.match(
      /<meta[^>]+content=["']([^"']+)["'][^>]+name=["']description["']/i,
    )?.[1] ||
    "";
  const ogDesc =
    html.match(
      /<meta[^>]+property=["']og:description["'][^>]+content=["']([^"']+)["']/i,
    )?.[1] || "";
  const h1 = html.match(/<h1[^>]*>([\s\S]*?)<\/h1>/i)?.[1] || "";
  const h2s = [...html.matchAll(/<h2[^>]*>([\s\S]*?)<\/h2>/gi)]
    .slice(0, 3)
    .map((m) => m[1]);
  const ps = [...html.matchAll(/<p[^>]*>([\s\S]*?)<\/p>/gi)]
    .slice(0, 5)
    .map((m) => m[1]);

  const parts = [
    title && `TITLE: ${stripTags(title)}`,
    metaDesc && `DESC: ${stripTags(metaDesc)}`,
    ogDesc && metaDesc !== ogDesc && `OG: ${stripTags(ogDesc)}`,
    h1 && `H1: ${stripTags(h1)}`,
    h2s.length && `H2: ${h2s.map(stripTags).filter(Boolean).join(" | ")}`,
    ps.length && `P: ${ps.map(stripTags).filter(Boolean).join(" · ")}`,
  ].filter(Boolean);

  const signals = parts.join("\n").slice(0, MAX_SIGNAL_CHARS);
  return signals;
}

const SYSTEM = `Você ajuda a personalizar emails de cold outreach em português brasileiro.

Receberá: o nome de uma empresa e trechos do site oficial dela (título, meta, h1, parágrafos).

Sua tarefa: gerar um JSON com dois campos:
- "subject": linha de assunto curta (máx 55 caracteres), específica pro nicho da empresa, sem clickbait, sem caps lock, sem emoji, sem palavras de spam ("grátis", "oportunidade", "urgente"). Pode ser em letras minúsculas, estilo conversacional.
- "hook": 1 frase (máx 220 caracteres) de abertura natural que referencia ESPECIFICAMENTE o que a empresa faz com base nos trechos fornecidos. Tom casual, direto, em primeira pessoa ("vi que vocês..."). Sem elogio genérico ("site bonito"), sem inventar fatos não presentes nos trechos, sem prêmios/clientes não citados.

Regras importantes:
- Se os trechos forem vagos ou insuficientes, escreva um hook neutro mas honesto baseado só no nicho aparente (não invente nada).
- Nunca mencione monitoramento, uptime, SaaS, UpStat ou qualquer produto — isso já está no corpo do email. O hook é só a abertura.
- Responda APENAS com JSON válido, sem markdown, sem comentários. Exemplo: {"subject":"...","hook":"..."}`;

function buildUserPrompt(name, signals) {
  return `Empresa: ${name}

Trechos do site:
${signals || "(site não retornou conteúdo útil)"}

Gere o JSON.`;
}

function sanitize(str, maxLen) {
  return String(str || "")
    .replace(/\s+/g, " ")
    .replace(/^["'`\s]+|["'`\s]+$/g, "")
    .slice(0, maxLen)
    .trim();
}

function parseResponse(content) {
  if (!content) return null;
  const jsonMatch = content.match(/\{[\s\S]*\}/);
  if (!jsonMatch) return null;
  try {
    const obj = JSON.parse(jsonMatch[0]);
    const subject = sanitize(obj.subject, 70);
    const hook = sanitize(obj.hook, 260);
    if (!subject || !hook) return null;
    return { subject, hook };
  } catch {
    return null;
  }
}

let client;
function getClient() {
  if (!process.env.GROQ_API_KEY)
    throw new Error("GROQ_API_KEY ausente no .env");
  if (!client) client = new Groq({ apiKey: process.env.GROQ_API_KEY });
  return client;
}

export async function personalizeLead(lead) {
  const groq = getClient();
  let signals = "";
  if (lead.website) {
    const html = await fetchHtml(lead.website);
    if (html) signals = extractSignals(html);
  }

  const completion = await groq.chat.completions.create({
    model: MODEL,
    temperature: 0.7,
    max_tokens: 300,
    response_format: { type: "json_object" },
    messages: [
      { role: "system", content: SYSTEM },
      { role: "user", content: buildUserPrompt(lead.name, signals) },
    ],
  });

  const content = completion.choices?.[0]?.message?.content || "";
  const parsed = parseResponse(content);
  if (!parsed) throw new Error("resposta inválida do Groq");
  return parsed;
}

export async function personalizeLeads(
  leads,
  { force = false, onProgress = consoleProgress } = {},
) {
  const targets = leads.filter(
    (l) => force || !(l.personalizedHook && l.personalizedSubject),
  );
  if (targets.length === 0) {
    onProgress({ type: "done", message: "Nada pra personalizar." });
    return { ok: 0, fail: 0 };
  }

  onProgress({
    type: "log",
    message: `Personalizando ${targets.length} leads via Groq (${MODEL})…`,
  });

  let ok = 0,
    fail = 0;
  for (let i = 0; i < targets.length; i++) {
    const lead = targets[i];
    try {
      const { subject, hook } = await personalizeLead(lead);
      lead.personalizedSubject = subject;
      lead.personalizedHook = hook;
      lead.personalizedAt = new Date().toISOString();
      ok++;
      onProgress({
        type: "item",
        index: i + 1,
        total: targets.length,
        name: lead.name,
        status: `ok · "${subject.slice(0, 40)}…"`,
      });
    } catch (e) {
      fail++;
      onProgress({
        type: "item",
        index: i + 1,
        total: targets.length,
        name: lead.name,
        status: `falhou: ${e.message}`,
      });
    }
  }
  onProgress({
    type: "done",
    message: `Personalização: ${ok} ok, ${fail} falhas.`,
    ok,
    fail,
  });
  return { ok, fail };
}

File: python/personalize.py
"""Geração de hook personalizado via Groq — equivalente a lib/personalize.js."""

import json
import os
import re
from datetime import datetime, timezone

import httpx
from groq import Groq

from progress import console_progress

MODEL = "llama-3.3-70b-versatile"
FETCH_TIMEOUT_S = 8.0
MAX_SIGNAL_CHARS = 1800

_HEADERS = {
    "User-Agent": (
        "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 "
        "(KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36"
    ),
    "Accept-Language": "pt-BR,pt;q=0.9,en;q=0.5",
}

SYSTEM = """Você ajuda a personalizar emails de cold outreach em português brasileiro.

Receberá: o nome de uma empresa, trechos do site oficial dela e, quando existir, uma análise técnica simples do site.

Sua tarefa: gerar um JSON com um campo:
- "hook": 1 frase (máx 220 caracteres) de abertura natural que referencia ESPECIFICAMENTE o que a empresa faz com base nos trechos fornecidos. Tom casual, direto, em primeira pessoa ("vi que vocês..."). Sem elogio genérico ("site bonito"), sem inventar fatos não presentes nos trechos, sem prêmios/clientes não citados.

Regras importantes:
- Se os trechos forem vagos ou insuficientes, escreva um hook neutro mas honesto baseado só no nicho aparente (não invente nada).
- Se houver análise técnica, use apenas como contexto secundário. Prefira mencionar serviços reais da empresa; só mencione stack/dor se isso estiver claro e não soar acusatório.
- Nunca mencione monitoramento, uptime, SaaS, UpStat ou qualquer produto — isso já está no corpo do email. O hook é só a abertura.
- Responda APENAS com JSON válido, sem markdown, sem comentários. Exemplo: {"hook":"..."}"""


def fetch_html(url):
    try:
        with httpx.Client(
            timeout=FETCH_TIMEOUT_S, follow_redirects=True, headers=_HEADERS
        ) as client:
            res = client.get(url)
        if not res.is_success:
            return ""
        ct = res.headers.get("content-type", "")
        if "text" not in ct and "html" not in ct:
            return ""
        return res.text
    except Exception:
        return ""


def strip_tags(s):
    s = re.sub(r"<script[\s\S]*?</script>", " ", s, flags=re.IGNORECASE)
    s = re.sub(r"<style[\s\S]*?</style>", " ", s, flags=re.IGNORECASE)
    s = re.sub(r"<[^>]+>", " ", s)
    s = re.sub(r"&nbsp;", " ", s, flags=re.IGNORECASE)
    s = re.sub(r"&amp;", "&", s, flags=re.IGNORECASE)
    s = re.sub(r"\s+", " ", s)
    return s.strip()


def _first(pattern, html, flags=re.IGNORECASE):
    m = re.search(pattern, html, flags)
    return m.group(1) if m else ""


def extract_signals(html):
    title = _first(r"<title[^>]*>([\s\S]*?)</title>", html)
    meta_desc = _first(
        r"<meta[^>]+name=[\"']description[\"'][^>]+content=[\"']([^\"']+)[\"']", html
    ) or _first(
        r"<meta[^>]+content=[\"']([^\"']+)[\"'][^>]+name=[\"']description[\"']", html
    )
    og_desc = _first(
        r"<meta[^>]+property=[\"']og:description[\"'][^>]+content=[\"']([^\"']+)[\"']",
        html,
    )
    h1 = _first(r"<h1[^>]*>([\s\S]*?)</h1>", html)
    h2s = [m.group(1) for m in re.finditer(r"<h2[^>]*>([\s\S]*?)</h2>", html, re.IGNORECASE)][:3]
    ps = [m.group(1) for m in re.finditer(r"<p[^>]*>([\s\S]*?)</p>", html, re.IGNORECASE)][:5]

    parts = []
    if title:
        parts.append(f"TITLE: {strip_tags(title)}")
    if meta_desc:
        parts.append(f"DESC: {strip_tags(meta_desc)}")
    if og_desc and meta_desc != og_desc:
        parts.append(f"OG: {strip_tags(og_desc)}")
    if h1:
        parts.append(f"H1: {strip_tags(h1)}")
    if h2s:
        parts.append("H2: " + " | ".join(filter(None, (strip_tags(x) for x in h2s))))
    if ps:
        parts.append("P: " + " · ".join(filter(None, (strip_tags(x) for x in ps))))

    return "\n".join(parts)[:MAX_SIGNAL_CHARS]


def format_insights(insights):
    if not insights:
        return ""
    stack = (
        f"STACK: {', '.join(insights['techStack'])}"
        if insights.get("techStack")
        else ""
    )
    pains = ""
    if insights.get("painSignals"):
        pains = "SINAIS: " + " | ".join(
            f"{s['label']} ({s['detail']})" for s in insights["painSignals"][:4]
        )
    pages = ""
    if insights.get("pages"):
        ok_pages = [p["path"] for p in insights["pages"] if p.get("ok")][:5]
        if ok_pages:
            pages = "PÁGINAS: " + ", ".join(ok_pages)
    return "\n".join(filter(None, [stack, pains, pages]))


def build_user_prompt(name, signals, insights):
    tech_signals = format_insights(insights)
    return f"""Empresa: {name}

Trechos do site:
{signals or "(site não retornou conteúdo útil)"}

Análise técnica:
{tech_signals or "(sem análise técnica salva)"}

Gere o JSON."""


def sanitize(s, max_len):
    s = str(s or "")
    s = re.sub(r"\s+", " ", s)
    s = re.sub(r"^[\"'`\s]+|[\"'`\s]+$", "", s)
    return s[:max_len].strip()


def parse_response(content):
    if not content:
        return None
    m = re.search(r"\{[\s\S]*\}", content)
    if not m:
        return None
    try:
        obj = json.loads(m.group(0))
        hook = sanitize(obj.get("hook"), 260)
        if not hook:
            return None
        return {"hook": hook}
    except Exception:
        return None


_client = None


def get_client():
    global _client
    if not os.environ.get("GROQ_API_KEY"):
        raise RuntimeError("GROQ_API_KEY ausente no .env")
    if _client is None:
        _client = Groq(api_key=os.environ["GROQ_API_KEY"])
    return _client


def personalize_lead(lead):
    groq = get_client()
    signals = ""
    if lead.get("website"):
        html = fetch_html(lead["website"])
        if html:
            signals = extract_signals(html)

    completion = groq.chat.completions.create(
        model=MODEL,
        temperature=0.7,
        max_tokens=300,
        response_format={"type": "json_object"},
        messages=[
            {"role": "system", "content": SYSTEM},
            {
                "role": "user",
                "content": build_user_prompt(
                    lead.get("name"), signals, lead.get("siteInsights")
                ),
            },
        ],
    )

    content = ""
    if completion.choices:
        content = completion.choices[0].message.content or ""
    parsed = parse_response(content)
    if not parsed:
        raise RuntimeError("resposta inválida do Groq")
    return parsed


def personalize_leads(leads, force=False, on_progress=console_progress):
    targets = [l for l in leads if force or not l.get("personalizedHook")]
    if not targets:
        on_progress({"type": "done", "message": "Nada pra personalizar."})
        return {"ok": 0, "fail": 0}

    on_progress(
        {
            "type": "log",
            "message": f"Personalizando {len(targets)} leads via Groq ({MODEL})…",
        }
    )

    ok = 0
    fail = 0
    for i, lead in enumerate(targets):
        try:
            hook = personalize_lead(lead)["hook"]
            lead["personalizedHook"] = hook
            lead["personalizedAt"] = datetime.now(timezone.utc).isoformat()
            ok += 1
            on_progress(
                {
                    "type": "item",
                    "index": i + 1,
                    "total": len(targets),
                    "name": lead.get("name"),
                    "status": f'ok · "{hook[:40]}…"',
                }
            )
        except Exception as e:
            fail += 1
            on_progress(
                {
                    "type": "item",
                    "index": i + 1,
                    "total": len(targets),
                    "name": lead.get("name"),
                    "status": f"falhou: {e}",
                }
            )
    on_progress(
        {
            "type": "done",
            "message": f"Personalização: {ok} ok, {fail} falhas.",
            "ok": ok,
            "fail": fail,
        }
    )
    return {"ok": ok, "fail": fail}
